src/python/pants/core/goals/generate_lockfiles.py

```diff
@@ -131,22 +133,76 @@ def __init__(
         )
 
 
+class _ResolveProviderType(Enum):
+    TOOL = 1
+    USER = 2
+
+
+@dataclass(frozen=True, order=True)
+class _ResolveProvider:
+    option_name: str
+    type_: _ResolveProviderType
+
+
 class AmbiguousResolveNamesError(Exception):
-    def __init__(self, ambiguous_names: list[str]) -> None:
-        if len(ambiguous_names) == 1:
-            first_paragraph = (
-                "A resolve name from the option `[python].experimental_resolves` collides with the "
-                f"name of a tool resolve: {ambiguous_names[0]}"
-            )
+    def __init__(self, ambiguous_name: str, providers: set[_ResolveProvider]) -> None:
+        tool_providers = []
+        user_providers = []
+        for provider in sorted(providers):
+            if provider.type_ == _ResolveProviderType.TOOL:
+                tool_providers.append(provider.option_name)
+            else:
+                user_providers.append(provider.option_name)
+
+        if tool_providers:
+            if not user_providers:
+                raise AssertionError(
+                    f"Two tools have the same options_scope: {ambiguous_name}. This "
+                    "should not be possible. Please open a bug at "
+                    "https://github.com/pantsbuild/pants/issues/new."
+                )
```

Comment:
Since users can also write plugins, it might be better to suggest that it is a conflict between plugins (in case they are actively editing).